### ISSUE TITLE
Resize transform produces aliasing artifacts

### ISSUE THREAD

**Issue by lemairecarl** [Project-MONAI/MONAI#4020]
**Describe the bug**
The Resize transform produces aliasing artifacts. It uses the `F.interpolate` function from PyTorch, which has an antialiasing option, but that does not support 3D downsampling of volumes (5D tensors). The Resize transforms does not use the antialiasing option at all: https://github.com/Project-MONAI/MONAI/blob/a676e3876903e5799eb3d24245872ea4f3a5b152/monai/transforms/spatial/array.py#L676

This line in PyTorch source code confirms that the antialiasing option is not available for 3D volumes (trilinear interpolation):
https://github.com/pytorch/pytorch/blob/8ed6cb42ba2635319b67417b9ee2f1a0142d2c25/torch/nn/functional.py#L3851

The issue of ignoring aliasing in deep learning research is covered in this paper: https://arxiv.org/abs/2104.11222

**To Reproduce**
1. Resize an image or volume down to a small size, e.g. using a factor of 8
2. Observe the aliasing artifacts in the resized tensor

**Expected behavior**
The downsampled image should look like a smaller version of the original (no aliasing).

**Comment by Nic-Ma** [comment]
Hi @lemairecarl ,

Thanks for your finding and feedback!
@wyli , I think maybe it's necessary to add the `antialias` arg to `Resize`, `Zoom`, `RandZoom` transforms based on:
https://github.com/pytorch/pytorch/blob/v1.11.0/torch/nn/functional.py#L3780
What do you think?

Thanks.

**Comment by wyli** [comment]
Yes there's also a related p1 milestone but I don't have the bandwidth to implement it https://github.com/Project-MONAI/MONAI/milestone/44

**Comment by Nic-Ma** [comment]
Hi @wyli ,

OK, as you and @yiheng-wang-nv are busy with other bugs, let me self-assign this ticket later.

Thanks.

**Comment by lemairecarl** [comment]
Hi, thanks for taking this seriously.

I see that you added this to the "label downsampling with antialiasing" milestone. Maybe you should rename the milestone as just "downsampling with antialiasing" ?

**Comment by dongyang0122** [comment]
hi @lemairecarl, thanks for reporting the issue! Do you have sampling images (before and after the down-sampling) showing the undesired artifacts? That would be helpful for us to understand and resolve the issue.

**Comment by wyli** [comment]
> hi @lemairecarl, thanks for reporting the issue! Do you have sampling images (before and after the down-sampling) showing the undesired artifacts? That would be helpful for us to understand and resolve the issue.

Hi @dongyang0122 the appendix section 6.4 figure 12 of the paper is a good demo https://arxiv.org/pdf/2104.11222.pdf

**Comment by lemairecarl** [comment]
I don't have an example on hand right now. I know that down-scaling using interpolation without applying a low-pass filter before will give a result that is similar to down-scaling using "nearest neighbor". You can try down-scaling an image with "nearest neighbor" in GIMP or Photoshop to get a sense of it. It introduces unwanted distortions and high frequencies. I suppose you can train a model that appears successful even if those distortions are present; but they should not be there.

**Comment by lemairecarl** [comment]
Here is an example:

![downsample_example](https://user-images.githubusercontent.com/13444373/162337362-433cfbe4-44a6-4528-a7fa-a6992a941b25.png)

I think one could argue that the aliased version is a reasonable representation. However, the anti-aliased version is more faithful. I would say that the blurriness is informative. For example, look at the center of the image, where the right structure penetrates the left one with a kind of "arm". In the aliased version, the arm is pretty much gone. In the anti-aliased version, the arm is still there -- or at least we know that something is there, even if blurry.

**Comment by lemairecarl** [comment]
It seems that in Pillow, the filtering is done at the same time as the resampling. Here is where a pixel value is multiplied with a filter value (`k[x]`):
https://github.com/python-pillow/Pillow/blob/63f64de8b5cc6d065f19cc73c865633ab4940a52/src/libImaging/Resample.c#L493

The filter used for bilinear filtering is defined here:
https://github.com/python-pillow/Pillow/blob/63f64de8b5cc6d065f19cc73c865633ab4940a52/src/libImaging/Resample.c#L21

Now, if we want to avoid implementing something in C++, one approach I see is the following:

1. Build a low-pass filter
2. Apply the low pass filter to the volume using `F.conv3d`
3. Call `F.interpolate`

---

Another option which I've just discovered is to use scipy's `affine_transform` (https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.affine_transform.html). By default, a spline filtering is applied. The downside of this function is that we have to build a transformation matrix.

**Comment by miccai-tutorials** [comment]
hi @lemairecarl, thank you for the example! We are currently working on solutions for the anti-aliasing issue. Adding low-pass filter is one of options. And such filtering will be implemented with torch tensor.

**Comment by lemairecarl** [comment]
@miccai-tutorials Awesome, thanks!

**Comment by vfdev-5** [comment]
@wyli by the way, is there any interest in enabling antialias option from pytorch side for 5D ?
I was implementing it for pytorch and main blocker was in backward pass that can't be easily generalized to Nd case.

**Comment by wyli** [comment]
yes, I think spatially 2d and 3d will be very helpful already. is the backward pass some non-trainable deconvolution?

**Comment by Can-Zhao** [comment]
> @wyli by the way, is there any interest in enabling antialias option from pytorch side for 5D ? I was implementing it for pytorch and main blocker was in backward pass that can't be easily generalized to Nd case.

Thanks. In https://github.com/Project-MONAI/MONAI/blob/7f80288381575121e670a4e184d45b01813a51be/monai/transforms/spatial/array.py#L760-L771 ,
you can see that there is no constraint in the spatial dimension of images for the anti-aliasing part...

**Comment by vfdev-5** [comment]
@Can-Zhao, yes, applying gaussian blur can be also an option. It is a two steps resizing, it would be nice to benchmark these two implementations.

**Comment by Can-Zhao** [comment]
> @Can-Zhao, yes, applying gaussian blur can be also an option. It is a two steps resizing, it would be nice to benchmark these two implementations.

Sorry I did not realize it goes to c++ level. I might misunderstand your concern...
